vhost/src/vhost_user/gpu_message.rs

```diff
@@ -295,6 +295,28 @@ unsafe impl ByteValued for VhostUserGpuDMABUFScanout {}
 
 impl VhostUserMsgValidator for VhostUserGpuDMABUFScanout {}
 
+/// Update the scanout content.
+/// Data is struct VhostUserGpuUpdate
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C)]
+pub struct VhostUserGpuUpdate {
+    /// scanout content to update
+    pub scanout_id: u32,
+    /// The position field x of the scanout update
+    pub x: u32,
+    /// The position field y of the scanout update
+    pub y: u32,
+    /// scanout updated width size
+    pub width: u32,
+    /// scanout updated height size
+    pub height: u32,
+    /// the RGB data
+    pub data: [u8; 8],
```

Comment:
The length 8 is not correct.
In C it is supposed to be a flexible array member [1] [2]. Rust also support a similar thing, you can find it under the name dynamically sized types [3], but the language support for this feature is not finished yet. So I think it is impossible to use here (especially with derive `Default` and unsafe impl `ByteValued`)

I suggest removing data field from this struct and accepting another argument `data: Vec<u8>` in your `get_update` method.
Then writing the data into the socket stream after this `VhostUserGpuUpdate`. 

[1] https://github.com/qemu/qemu/blob/ce64e6224affb8b4e4b019f76d2950270b391af5/hw/display/vhost-user-gpu.c#L67
[2] https://en.wikipedia.org/wiki/Flexible_array_member
[3] https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts

Reply:
I thought about even using `data: Vec<u8>` but in get_update method it requires that the trait `std::marker::Copy` be implemented for this type. and Vec<u8> does not implement `std::marker::Copy`

Reply:
Yes, putting a `Vec` into this struct doesn't help. (it's memory represention is not an array, but 2 ints and a pointer to an array).

You can remove the `data` field from this struct. Accept both ` VhostUserGpuUpdate` and `data: Vec<u8>` as arguments to `get_update` method. And then you have to write some custom logic into the method - because you need to do 2 writes to the socket. First write is the `VhostUserGpuUpdate` and the other write would be the bytes inside the `Vec<u8>`. (not the Vec itself, just the bytes inside it!). You might need to add another helper method, to write a `&[u8]` into the socket instead of something ByteValued.

Reply:
we could remove data field entirely, because I see it's not even used in QEMU.

Reply:
I didn't realize we had something similar in vhost, thanks for the pointer.